core-tests/shared/src/test/scala/zio/ZIOSpec.scala

```diff
@@ -2682,6 +2686,24 @@ object ZIOSpec extends ZIOBaseSpec {
 
         assertM(Live.live(io))(isTrue)
       },
+      testM("does not make effect interruptible") {
```

Comment:
What does this test? There is no uninterruptible region, all the effects in this test are interruptible.

There's a test for the behavior I'm talking about:

```scala
for {
    p <- Promise.make[String, Unit]
    fiber <- ZIO.never.disconnect.onInterrupt(p.fail("Interrupted even when inside uninterruptible region")).uninterruptible.fork
    res <- fiber.interrupt.timeout(5.seconds)
  } yield res.isEmpty // interruption shouldn't succeed
```

Here the changes to `disconnect` will force the fiber's `.onInterrupt` to happen despite the entire fiber being uninterruptible.

Reply:
The finalizer in `ensuring` will be uninterruptible. If I replace `disconnect` with `interruptible` there the test will fail.

Reply:
@adamgfraser The effect inside `.disconnect` will be uninterruptible inded, but everything outside, inside the `ensuring` will still be interrupted. Try with `ZIO.unit.disconnect *> latch2.succeed(())` - this should fail

Reply:
@neko-kai Yes I see your point. I am wondering whether there is a way to address both of these points. I think you're right that we shouldn't change the interruptibility of the outer effect. But the fact that we can't just use `disconnect` in situations like this and need to then make the disconnection interruptible seems problematic to me. But I may just be wrong there. 😃

Reply:
@adamgfraser I think it's ok to use `.disconnect.interruptible.ensuring(<rest-of-the-finalizer>)` explicitly in these situations, when you really need to disconnect early inside a finalizer. There's no way to react to interruption, without being interrupted - that means there's no way to be disconnected without being interrupted too. This tradeoff is fine by me, I think it's reasonable, I have done a lot of arguing to make sure that interruption is irreversible and intend to keep doing it. Everyone who knows what they're doing can achieve the  desired behavior explicitly, but we shouldn't open interruptibility holes in innocuous looking functions because this can seriously mess up with people's ability to understand what happens to their code.

Reply:
@neko-kai Yes. Updated to incorporate your feedback. Thank you for pushing here.